src/jit/importer.cpp

```diff
@@ -4045,6 +4092,125 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
     return retNode;
 }
 
+#ifdef FEATURE_HW_INTRINSICS
+//------------------------------------------------------------------------
+// impBaseIntrinsic: dispatch intrinsics to their own implementation
+//
+// Arguments:
+//    intrinsic  -- id of the intrinsic function.
+//    method     -- method handle of the intrinsic function.
+//    sig        -- signature of the intrinsic call
+//
+// Return Value:
+//    the expanded intrinsic.
+//
+GenTree* Compiler::impBaseIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* sig)
+{
+    GenTree*  retNode  = nullptr;
+    unsigned  simdSize = 0;
+    var_types baseType = getBaseTypeAndSizeOfSIMDType(sig->retTypeClass, &simdSize);
+    var_types retType  = getSIMDTypeForSize(simdSize);
+
+    if (sig->hasThis())
+    {
+        CORINFO_CLASS_HANDLE thisClass = info.compCompHnd->getArgClass(sig, sig->args);
+        var_types            thisType  = getBaseTypeOfSIMDType(thisClass);
+
+        if (!varTypeIsArithmetic(thisType))
+        {
+            return nullptr;
+        }
+    }
+
+    if (!varTypeIsArithmetic(baseType))
+    {
+        return nullptr;
+    }
+
+    switch (intrinsic)
+    {
+#if defined(_TARGET_ARM64_)
+        case NI_Base_Vector64_AsByte:
+        case NI_Base_Vector64_AsInt16:
+        case NI_Base_Vector64_AsInt32:
+        case NI_Base_Vector64_AsSByte:
+        case NI_Base_Vector64_AsSingle:
+        case NI_Base_Vector64_AsUInt16:
+        case NI_Base_Vector64_AsUInt32:
+#endif // _TARGET_ARM64_
+        case NI_Base_Vector128_As:
+        case NI_Base_Vector128_AsByte:
+        case NI_Base_Vector128_AsDouble:
+        case NI_Base_Vector128_AsInt16:
+        case NI_Base_Vector128_AsInt32:
+        case NI_Base_Vector128_AsInt64:
+        case NI_Base_Vector128_AsSByte:
+        case NI_Base_Vector128_AsSingle:
+        case NI_Base_Vector128_AsUInt16:
+        case NI_Base_Vector128_AsUInt32:
+        case NI_Base_Vector128_AsUInt64:
+#if defined(_TARGET_XARCH_)
+        case NI_Base_Vector256_As:
+        case NI_Base_Vector256_AsByte:
+        case NI_Base_Vector256_AsDouble:
+        case NI_Base_Vector256_AsInt16:
+        case NI_Base_Vector256_AsInt32:
+        case NI_Base_Vector256_AsInt64:
+        case NI_Base_Vector256_AsSByte:
+        case NI_Base_Vector256_AsSingle:
+        case NI_Base_Vector256_AsUInt16:
+        case NI_Base_Vector256_AsUInt32:
+        case NI_Base_Vector256_AsUInt64:
+#endif // _TARGET_XARCH_
+        {
+            // We fold away the cast here, as it only exists to satisfy
+            // the type system. It is safe to do this here since the retNode type
+            // and the signature return type are both the same TYP_SIMD.
```

Comment:
This is still a bit messy, because we can create values without a valid known handle (I've run into this a few times in trying to eliminate `OBJ` and `BLK` nodes above SIMD values). Eventually I'd like to see us eliminate reliance on handles for SIMD values at all. However, doing so requires, at a minimum, that we bake-in an understanding of the ABI handling for each SIMD type. There may be other issues lurking as well. For now, I think this works because we always create an `OBJ` node for args, and that carries the actual signature handle.

Reply:
Right - so I'm thinking that perhaps we need different types for, e.g. `Vector4` vs. `Vector128<T>` or `Vector<T>`, because the former is an actual struct of 4 fields while the other two are opaque vectors, and should be passed as a vector when supported by the ABI.

Reply:
Ah, I see.

Maybe this could be handled by getting `Vector4` updated to be implemented in terms of `Vector128<T>` (which was discussed in the past). This is similar to what most native code ends up doing

Reply:
I'm not sure whether `Vector<T>` should be considered an ABI type though.... I think it might be confusing given its dynamic size/etc (makes it harder to use as an interchange type, but easier to use as an implementation detail).

Reply:
Right - I don't see us ever making `Vector<T>` an interchange type - but if we had a vector ABI we'd want to pass it as a vector.

Implementing `Vector4` on top of `Vector128<T>` will require care to ensure that both correctness (still follows the "normal struct" ABI) and performance are preserved.

Reply:
> but if we had a vector ABI we'd want to pass it as a vector.

Right, but I think that might be able to just "fallout" from it being an aggregate type. The System V and Windows ABIs define cases for both `struct s { __m128 x; __m128 y; __m128 z; __m128 w; }` and `struct s { float x; float y; float z; float w; }`, for example.

> Implementing Vector4 on top of Vector128<T> will require care to ensure that both correctness (still follows the "normal struct" ABI) and performance are preserved.

Yes, ensuring the codegen remains good when converting between a `user-defined struct` and a `__m128` is important (and likely something we need to have handled anyways).